### PR TITLE
3070-3071 generic model program/instance resources migration

### PR DESCRIPTION
**Pull request by pkdash**
### Pull Request Checklist: 
- [x] Positive Test Case Written by Dev

- [x] Automated Testing

- [x] Sufficient User and Developer Documentation

- [x] Passing Jenkins Build

- [x] Peer Code review and approval

### Positive Test Case
**NOTE**: The model migration commands are only for migration model program resource and generic model instance resource to composite resource. Until we have the migration commands for migrating both MODFLOW and SWAT model instance resources as part of a release and deployment to production, the model migration commands described here should not be executed in production. 
 
Refer to this google document for rules and requirements for migration: https://docs.google.com/document/d/1gBa8VHB1Yb-Ub4iwHNTOSiblACrEODPvt3cZoCqQFp0/edit
 
1. The model resource migration commands need to be run against real resources for the purpose of testing.
2.  If you are using beta for this testing, then run the following  two commands to clean up folders that might have been created from previous test of model migration.

 `delete_model_aggregation_folder_from_irods ModelProgram`

The above command will delete folders created in iRODS as part of migrating model program resources.

`delete_model_aggregation_folder_from_irods ModelInstance`

The above command will delete folders created in iRODS as part of migrating generic model instance resources.

NOTE: Both these commands need to be executed.

3. Then run the following management command to prepare all model instance resources for migration to composite resource. This management command captures the resource id of any linked model program resource and stores in the model instance resource `extra_data` field later to be used when migrating model instance resources.

 `prepare_model_instance_resources_for_migration`

4. Run the following command to migrate all generic model program resources to composite resource.

`migrate_model_program_resources`

NOTE: The above command can take a while to finish. There are about 184 model program resources to be migrated. The last output from the above command should look like this:

**0 MODEL PROGRAM RESOURCES HAD ISSUES DURING MIGRATION TO COMPOSITE RESOURCE**
**184 MODEL PROGRAM RESOURCES WERE MIGRATED TO COMPOSITE RESOURCE**
**ALL MODEL PROGRAM RESOURCES WERE MIGRATED TO COMPOSITE RESOURCE**

5. Once the above model program resource migration finishes successfully (ALL MODEL PROGRAM RESOURCES WERE MIGRATED TO COMPOSITE RESOURCE), then run the following command to migrate all generic model instance resources.

`migrate_model_instance_resources`

NOTE: The above command can take a while to finish. There are about 265 model instance resources to be migrated. The last output from the command should look like this:

**0 MODEL INSTANCE RESOURCES HAD ISSUES DURING MIGRATION TO COMPOSITE RESOURCE**
**265 MODEL INSTANCE RESOURCES WERE MIGRATED TO COMPOSITE RESOURCE**
**ALL MODEL INSTANCE RESOURCES WERE MIGRATED TO COMPOSITE RESOURCE**

**Cleanup beta after testing model migration:**
1. Rebuild beta with data update
2.  Then run the following commands:

 `delete_model_aggregation_folder_from_irods ModelProgram`

 `delete_model_aggregation_folder_from_irods ModelInstance`

### CODE REVIEW

**Comment by hydrocheck** [comment]
| METRIC | VALUE |
|:-----:|:-----:|
| NEW SECURITY HOTSPOTS | 4 |
| NEW BUGS | 0 |
| NEW CODE SMELLS | 9 |
| NEW DUPLICATED LINES DENSITY | 25.0 |
| NEW VULNERABILITIES | 0 |

[https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4421](https://sonarqube.cuahsi-workstation.com:9000/dashboard?id=hydroshare-4421)

File: hs_core/management/commands/delete_model_aggregation_folder_from_irods.py [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+    def process_resource(self, model_res, resource_type, logger):
+        istorage = model_res.get_irods_storage()
+        model_res_type_folder_mapping = {"ModelProgram": "model-program", "ModelInstance": "model-instance",
+                                         "MODFLOWModelInstance": "modflow-instance",
+                                         "SWATModelInstance": "swat-instance"}
+        if not istorage.exists(model_res.root_path):
+            err_msg = ">> Resource (ID:{}) doesn't exist in iRODS"
+            self.stdout.write(self.style.ERROR(err_msg))
+            self.stdout.flush()
```
Comment (hyi) [inline hs_core/management/commands/delete_model_aggregation_folder_from_irods.py:35]: you can simplify it with ```print(message, flush=True)```

File: hs_core/tests/data/readme.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+readme
```
Comment (hyi) [inline hs_core/tests/data/readme.txt:1]: Is this a placeholder readme file or some brief real readme text needs to be put in the file?

File: hs_core/views/resource_folder_hierarchy.py [hunk 3]
```diff
@@ -158,7 +158,7 @@ def data_store_structure(request):
             if not main_extension:
                 # accept any extension
                 main_extension = ""
-            if main_extension.endswith(f.extension):
+            if main_extension.endswith(f.extension) and f.extension:
```
Comment (hyi) [inline hs_core/views/resource_folder_hierarchy.py:161]: Should the order be changed in the if statement, i.e., change it to ```if f.extension and main_extension.endswith(f.extension)```?

File: pylint.out [hunk 4]
```diff
@@ -0,0 +1,188 @@
+./hs_collection_resource/views.py:125:25: E117 over-indented
+./hs_collection_resource/patches/add_hasPart_term_django_shell.py:20:1: F841 local variable 'ex' is assigned to but never used
+./hs_file_types/nc_functions/nc_utils.py:292:61: W504 line break after binary operator
+./hs_file_types/nc_functions/nc_utils.py:354:49: E225 missing whitespace around operator
```
Comment (hyi) [inline pylint.out:4]: this pylint.out needs to be removed.

**Review by hyi — COMMENTED** [review]
@pkdash I have gone through the PR and left a couple of comments for your reference.

File: pylint.out (hunk 4 above)
Comment (pkdash) [reply pylint.out:4]: @hyi It seems this file currently exists in develop. May be we have to add this file to gitignore. I am going to make this change part of this PR unless we create a separate issue. Your thoughts?

File: hs_core/management/commands/delete_model_aggregation_folder_from_irods.py (hunk 1 above)
Comment (pkdash) [reply hs_core/management/commands/delete_model_aggregation_folder_from_irods.py:35]: I am using `self.style.ERROR(err_msg))`  to easily see the error messages (displayed in red text) when testing. That's the reason I didn't use `print`

File: hs_core/tests/data/readme.txt (hunk 2 above)
Comment (pkdash) [reply hs_core/tests/data/readme.txt:1]: The tests are relevant to the file name only (`readme.txt`) and not the content of this file.

File: hs_core/views/resource_folder_hierarchy.py (hunk 3 above)
Comment (pkdash) [reply hs_core/views/resource_folder_hierarchy.py]: Good catch. Fixed.

**Comment by pkdash** [comment]
@hyi See if I have addressed your comments.

File: pylint.out (hunk 4 above)
Comment (hyi) [reply pylint.out]: @pkdash Sounds like a good idea to add this file in gitignore. Agreed it is a good idea to address this in this PR.

**Review by hyi — APPROVED** [review]

**Pull request merged**